### PR TITLE
executor, types: refactor CompareDatum

### PR DESCRIPTION
**Pull request by wjhuang2016**
Relate to #29867
CompareDatum uses collation implicitly which tends to error.
So I add an argument collator explicitly.  Compare is introduced in this PR, and it will replace CompareDatum step by step.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -548,7 +549,63 @@ func (d *Datum) SetValue(val interface{}, tp *types.FieldType) {
 	}
 }
 
+// CompareDatumNew compares datum to another datum.
+// Notes: don't rely on datum.collation to get the collator, it's tend to buggy.
+// TODO: use this function to replace CompareDatum. After we remove all of usage of CompareDatum, we can rename this function back to CompareDatum.
+func (d *Datum) CompareDatumNew(sc *stmtctx.StatementContext, ad *Datum, comparer collate.Collator) (int, error) {
```
Comment (kennytm) [inline types/datum.go:555]: how about just call this `Compare()` 😐

Comment (wjhuang2016) [reply types/datum.go:555]: Do you mean `compare(a, b, collate)`?

Comment (kennytm) [reply types/datum.go:555]: yeah.

```suggestion
func (d *Datum) Compare(sc *stmtctx.StatementContext, ad *Datum, comparer collate.Collator) (int, error) {
```

**Review by xiongjiwei — APPROVED** [review]

**Review by kennytm — APPROVED** [review]

**Pull request merged**